File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/method/DropLastWhile.java
package joookiwi.collection.java.method;

import java.util.function.BooleanSupplier;
import java.util.function.Predicate;
import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.GenericCollectionHolder;
import joookiwi.collection.java.MinimalistCollectionHolder;
import joookiwi.collection.java.annotation.ExtensionFunction;
import joookiwi.collection.java.callback.ObjIntPredicate;
import joookiwi.collection.java.exception.ImpossibleConstructionException;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Unmodifiable;

import static joookiwi.collection.java.CollectionConstants.emptyArray;
import static joookiwi.collection.java.CollectionConstants.emptyCollectionHolder;
import static joookiwi.collection.java.CommonContracts.ALWAYS_FAIL_0;

@NotNullByDefault
public final class DropLastWhile
        extends Utility {

    @Contract(ALWAYS_FAIL_0)
    private DropLastWhile() { throw new ImpossibleConstructionException("The utility class “DropLastWhile” cannot be constructed.", DropLastWhile.class); }

    //#region -------------------- Facade methods --------------------

    //#region -------------------- predicate (T, int) → boolean --------------------

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                 final ObjIntPredicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.size();
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with2Argument(collection, predicate, size));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable CollectionHolder<? extends T> collection,
                                                                                 final ObjIntPredicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();
        if (collection.isEmpty())
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with2Argument(collection, predicate, collection.size()));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final T @Nullable @Unmodifiable [] collection,
                                                                                 final ObjIntPredicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.length;
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with2Argument(collection, predicate, size));
    }

    //#endregion -------------------- predicate (T, int) → boolean --------------------
    //#region -------------------- predicate (T) → boolean --------------------

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                 final Predicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.size();
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with1Argument(collection, predicate, size));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable CollectionHolder<? extends T> collection,
                                                                                 final Predicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();
        if (collection.isEmpty())
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with1Argument(collection, predicate, collection.size()));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final T @Nullable @Unmodifiable [] collection,
                                                                                 final Predicate<? super T> predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.length;
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with1Argument(collection, predicate, size));
    }

    //#endregion -------------------- predicate (T) → boolean --------------------
    //#region -------------------- predicate () → boolean --------------------

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][MinimalistCollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable MinimalistCollectionHolder<? extends T> collection,
                                                                                 final BooleanSupplier predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.size();
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with0Argument(collection, predicate, size));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] [collection][CollectionHolder]
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final @Nullable CollectionHolder<? extends T> collection,
                                                                                 final BooleanSupplier predicate) {
        if (collection == null)
            return emptyCollectionHolder();
        if (collection.isEmpty())
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with0Argument(collection, predicate, collection.size()));
    }

    /// Get a new [CollectionHolder] **not** having the last elements satisfying the given `predicate`
    ///
    /// @param collection The [nullable][Nullable] collection
    /// @param predicate  The given predicate
    /// @param <T>        The `collection` type
    /// @see <a href="https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html">Kotlin dropLastWhile(predicate)</a>
    @ExtensionFunction
    public static <T extends @Nullable Object> CollectionHolder<T> dropLastWhile(final T @Nullable @Unmodifiable [] collection,
                                                                                 final BooleanSupplier predicate) {
        if (collection == null)
            return emptyCollectionHolder();

        final var size = collection.length;
        if (size == 0)
            return emptyCollectionHolder();
        return new GenericCollectionHolder<>(__with0Argument(collection, predicate, size));
    }

    //#endregion -------------------- predicate () → boolean --------------------

    //#endregion -------------------- Facade methods --------------------
    //#region -------------------- Loop methods --------------------

    private static <T extends @Nullable Object> T @Unmodifiable [] __with0Argument(final MinimalistCollectionHolder<? extends T> collection,
                                                                                   final BooleanSupplier predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0)
            if (!predicate.getAsBoolean()) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                var index2 = -1;
                while (++index2 <= index)
                    newArrayFrom0ToIndex[index2] = collection.get(index2);
                return newArrayFrom0ToIndex;
            }
        return emptyArray();
    }

    private static <T extends @Nullable Object> T @Unmodifiable [] __with0Argument(final T @Unmodifiable [] collection,
                                                                                   final BooleanSupplier predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0)
            if (!predicate.getAsBoolean()) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                var index2 = -1;
                while (++index2 <= index)
                    newArrayFrom0ToIndex[index2] = collection[index2];
                return newArrayFrom0ToIndex;
            }
        return emptyArray();
    }


    private static <T extends @Nullable Object> T @Unmodifiable [] __with1Argument(final MinimalistCollectionHolder<? extends T> collection,
                                                                                   final Predicate<? super T> predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0) {
            final var value = collection.get(index);
            if (!predicate.test(value)) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                newArrayFrom0ToIndex[index] = value;
                var index2 = -1;
                while (++index2 < index)
                    newArrayFrom0ToIndex[index2] = collection.get(index2);
                return newArrayFrom0ToIndex;
            }
        }
        return emptyArray();
    }

    private static <T extends @Nullable Object> T @Unmodifiable [] __with1Argument(final T @Unmodifiable [] collection,
                                                                                   final Predicate<? super T> predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0) {
            final var value = collection[index];
            if (!predicate.test(value)) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                newArrayFrom0ToIndex[index] = value;
                var index2 = -1;
                while (++index2 < index)
                    newArrayFrom0ToIndex[index2] = collection[index2];
                return newArrayFrom0ToIndex;
            }
        }
        return emptyArray();
    }


    private static <T extends @Nullable Object> T @Unmodifiable [] __with2Argument(final MinimalistCollectionHolder<? extends T> collection,
                                                                                   final ObjIntPredicate<? super T> predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0) {
            final var value = collection.get(index);
            if (!predicate.test(value, index)) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                newArrayFrom0ToIndex[index] = value;
                var index2 = -1;
                while (++index2 < index)
                    newArrayFrom0ToIndex[index2] = collection.get(index2);
                return newArrayFrom0ToIndex;
            }
        }
        return emptyArray();
    }

    private static <T extends @Nullable Object> T @Unmodifiable [] __with2Argument(final T @Unmodifiable [] collection,
                                                                                   final ObjIntPredicate<? super T> predicate,
                                                                                   final int size) {
        var index = size;
        while (--index >= 0) {
            final var value = collection[index];
            if (!predicate.test(value, index)) {
                @SuppressWarnings("unchecked cast") final var newArrayFrom0ToIndex = (T[]) new Object[index + 1];
                newArrayFrom0ToIndex[index] = value;
                var index2 = -1;
                while (++index2 < index)
                    newArrayFrom0ToIndex[index2] = collection[index2];
                return newArrayFrom0ToIndex;
            }
        }
        return emptyArray();
    }

    //#endregion -------------------- Loop methods --------------------

}

File: javascript/src/method/dropLastWhile.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {Nullable} from "@joookiwi/type"

import type {CollectionHolder}                           from "../CollectionHolder"
import type {MinimalistCollectionHolder}                 from "../MinimalistCollectionHolder"
import type {BooleanCallback, RestrainedBooleanCallback} from "../type/callback"

import {CollectionConstants}           from "../CollectionConstants"
import {isArray}                       from "./isArray"
import {isArrayByStructure}            from "./isArrayByStructure"
import {isCollectionHolder}            from "./isCollectionHolder"
import {isCollectionHolderByStructure} from "./isCollectionHolderByStructure"
import {isMinimalistCollectionHolder}  from "./isMinimalistCollectionHolder"

//#region -------------------- Facade method --------------------

/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The {@link Nullable nullable} collection ({@link CollectionHolder}, {@link MinimalistCollectionHolder} or {@link ReadonlyArray Array})
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 * @typescriptDefinition
 */
export function dropLastWhile<const T, const S extends T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, predicate: RestrainedBooleanCallback<T, S>,): CollectionHolder<S>
/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The {@link Nullable nullable} collection ({@link CollectionHolder}, {@link MinimalistCollectionHolder} or {@link ReadonlyArray Array})
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 */
export function dropLastWhile<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, predicate: BooleanCallback<T>,): CollectionHolder<T>
export function dropLastWhile<const T, >(collection: Nullable<| MinimalistCollectionHolder<T> | CollectionHolder<T> | readonly T[]>, predicate: BooleanCallback<T>,) {
    if (collection == null)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER
    if (isCollectionHolder(collection,))
        return dropLastWhileByCollectionHolder(collection, predicate,)
    if (isArray(collection,))
        return dropLastWhileByArray(collection, predicate,)
    if (isMinimalistCollectionHolder(collection,))
        return dropLastWhileByMinimalistCollectionHolder(collection, predicate,)

    if (isCollectionHolderByStructure<T>(collection,))
        return dropLastWhileByCollectionHolder(collection, predicate,)
    if (isArrayByStructure<T>(collection,))
        return dropLastWhileByArray(collection, predicate,)
    return dropLastWhileByMinimalistCollectionHolder(collection, predicate,)
}


/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 * @typescriptDefinition
 */
export function dropLastWhileByMinimalistCollectionHolder<const T, const S extends T, >(collection: Nullable<MinimalistCollectionHolder<T>>, predicate: RestrainedBooleanCallback<T, S>,): CollectionHolder<S>
/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 */
export function dropLastWhileByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, predicate: BooleanCallback<T>,): CollectionHolder<T>
export function dropLastWhileByMinimalistCollectionHolder<const T, >(collection: Nullable<MinimalistCollectionHolder<T>>, predicate: BooleanCallback<T>,) {
    if (collection == null)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER

    const size = collection.size
    if (size === 0)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER
    if (predicate.length == 1)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with1Argument(collection, size, predicate as (value: T,) => boolean,),)
    if (predicate.length >= 2)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with2Argument(collection, size, predicate,),)
    return new CollectionConstants.LazyGenericCollectionHolder(() => __with0Argument(collection, size, predicate as () => boolean,),)
}

/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 * @typescriptDefinition
 */
export function dropLastWhileByCollectionHolder<const T, const S extends T, >(collection: Nullable<CollectionHolder<T>>, predicate: RestrainedBooleanCallback<T, S>,): CollectionHolder<S>
/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 */
export function dropLastWhileByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, predicate: BooleanCallback<T>,): CollectionHolder<T>
export function dropLastWhileByCollectionHolder<const T, >(collection: Nullable<CollectionHolder<T>>, predicate: BooleanCallback<T>,) {
    if (collection == null)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER
    if (collection.isEmpty)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER
    if (predicate.length == 1)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with1Argument(collection, collection.size, predicate as (value: T,) => boolean,),)
    if (predicate.length >= 2)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with2Argument(collection, collection.size, predicate,),)
    return new CollectionConstants.LazyGenericCollectionHolder(() => __with0Argument(collection, collection.size, predicate as () => boolean,),)
}

/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 * @typescriptDefinition
 */
export function dropLastWhileByArray<const T, const S extends T, >(collection: Nullable<readonly T[]>, predicate: RestrainedBooleanCallback<T, S>,): CollectionHolder<S>
/**
 * Get a new {@link CollectionHolder} <b>not</b> having the last elements satisfying the given {@link predicate}
 *
 * @param collection The nullable collection
 * @param predicate  The given predicate
 * @see https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/drop-last-while.html Kotlin dropLastWhile(predicate)
 */
export function dropLastWhileByArray<const T, >(collection: Nullable<readonly T[]>, predicate: BooleanCallback<T>,): CollectionHolder<T>
export function dropLastWhileByArray<const T, >(collection: Nullable<readonly T[]>, predicate: BooleanCallback<T>,) {
    if (collection == null)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER

    const size = collection.length
    if (size === 0)
        return CollectionConstants.EMPTY_COLLECTION_HOLDER
    if (predicate.length == 1)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with1ArgumentByArray(collection, size, predicate as (value: T,) => boolean,),)
    if (predicate.length >= 2)
        return new CollectionConstants.LazyGenericCollectionHolder(() => __with2ArgumentByArray(collection, size, predicate,),)
    return new CollectionConstants.LazyGenericCollectionHolder(() => __with0ArgumentByArray(collection, size, predicate as () => boolean,),)
}

//#endregion -------------------- Facade method --------------------
//#region -------------------- Loop methods --------------------

function __with0Argument<const T, >(collection: MinimalistCollectionHolder<T>, size: number, predicate: () => boolean,): readonly T[] {
    let index = size
    while (--index >= 0)
        if (!predicate()) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            let index2 = -1
            while (++index2 <= index)
                newArrayFrom0ToIndex[index2] = collection.get(index2,)
            return newArrayFrom0ToIndex
        }
    return CollectionConstants.EMPTY_ARRAY
}

function __with0ArgumentByArray<const T, >(collection: readonly T[], size: number, predicate: () => boolean,): readonly T[] {
    let index = size
    while (--index >= 0)
        if (!predicate()) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            let index2 = -1
            while (++index2 <= index)
                newArrayFrom0ToIndex[index2] = collection[index2] as T
            return newArrayFrom0ToIndex
        }
    return CollectionConstants.EMPTY_ARRAY
}


function __with1Argument<const T, >(collection: MinimalistCollectionHolder<T>, size: number, predicate: (value: T,) => boolean,): readonly T[] {
    let index = size
    while (--index >= 0) {
        const value = collection.get(index,)
        if (!predicate(value,)) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            newArrayFrom0ToIndex[index] = value
            let index2 = -1
            while (++index2 < index)
                newArrayFrom0ToIndex[index2] = collection.get(index2,)
            return newArrayFrom0ToIndex
        }
    }
    return CollectionConstants.EMPTY_ARRAY
}

function __with1ArgumentByArray<const T, >(collection: readonly T[], size: number, predicate: (value: T,) => boolean,): readonly T[] {
    let index = size
    while (--index >= 0) {
        const value = collection[index] as T
        if (!predicate(value,)) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            newArrayFrom0ToIndex[index] = value
            let index2 = -1
            while (++index2 < index)
                newArrayFrom0ToIndex[index2] = collection[index2] as T
            return newArrayFrom0ToIndex
        }
    }
    return CollectionConstants.EMPTY_ARRAY
}


function __with2Argument<const T, >(collection: MinimalistCollectionHolder<T>, size: number, predicate: (value: T, index: number,) => boolean,): readonly T[] {
    let index = size
    while (--index >= 0) {
        const value = collection.get(index,)
        if (!predicate(value, index,)) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            newArrayFrom0ToIndex[index] = value
            let index2 = -1
            while (++index2 < index)
                newArrayFrom0ToIndex[index2] = collection.get(index2,)
            return newArrayFrom0ToIndex
        }
    }
    return CollectionConstants.EMPTY_ARRAY
}

function __with2ArgumentByArray<const T, >(collection: readonly T[], size: number, predicate: (value: T, index: number,) => boolean,): readonly T[] {
    let index = size
    while (--index >= 0) {
        const value = collection[index] as T
        if (!predicate(value, index,)) {
            const newArrayFrom0ToIndex = new Array<T>(index + 1,)
            newArrayFrom0ToIndex[index] = value
            let index2 = -1
            while (++index2 < index)
                newArrayFrom0ToIndex[index2] = collection[index2] as T
            return newArrayFrom0ToIndex
        }
    }
    return CollectionConstants.EMPTY_ARRAY
}

//#endregion -------------------- Loop methods --------------------
